openpgp/v2/read.go

```diff
@@ -740,43 +740,76 @@ func verifyDetachedSignatureReader(keyring KeyRing, signed, signature io.Reader,
 }
 
 // checkSignatureDetails verifies the metadata of the signature.
-// Checks the following:
-// - Hash function should not be invalid.
+// It checks the following:
+// - Hash function should not be invalid according to
+//   config.RejectHashAlgorithms.
 // - Verification key must be older than the signature creation time.
 // - Check signature notations.
-// - Signature is not expired.
-func checkSignatureDetails(verifiedKey *Key, signature *packet.Signature, config *packet.Config) error {
-	var collectedErrors []error
-	now := config.Now()
+// - Signature is not expired (unless a zero time is passed to
+//   explicitly ignore expiration).
+func checkSignatureDetails(pk *packet.PublicKey, signature *packet.Signature, now time.Time, config *packet.Config) error {
+	if config.RejectHashAlgorithm(signature.Hash) {
+		return errors.SignatureError("insecure hash algorithm: " + signature.Hash.String())
+	}
 
-	if config.RejectMessageHashAlgorithm(signature.Hash) {
-		return errors.SignatureError("insecure message hash algorithm: " + signature.Hash.String())
+	if pk.CreationTime.Unix() > signature.CreationTime.Unix() {
+		return errors.ErrSignatureOlderThanKey
 	}
 
-	if verifiedKey.PublicKey.CreationTime.Unix() > signature.CreationTime.Unix() {
-		collectedErrors = append(collectedErrors, errors.ErrSignatureOlderThanKey)
+	for _, notation := range signature.Notations {
+		if notation.IsCritical && !config.KnownNotation(notation.Name) {
+			return errors.SignatureError("unknown critical notation: " + notation.Name)
+		}
 	}
 
-	sigsToCheck := []*packet.Signature{signature, verifiedKey.PrimarySelfSignature}
+	if !now.IsZero() && signature.SigExpired(now) {
+		return errors.ErrSignatureExpired
+	}
+	return nil
+}
 
+// checkMessageSignatureDetails verifies the metadata of the signature.
+// It checks the criteria of checkSignatureDetails for the message
+// signature and all relevant binding signatures.
+// In addition, the message signature hash algorithm is checked against
+// config.RejectMessageHashAlgorithms.
+func checkMessageSignatureDetails(verifiedKey *Key, signature *packet.Signature, config *packet.Config) error {
+	if config.RejectMessageHashAlgorithm(signature.Hash) {
+		return errors.SignatureError("insecure message hash algorithm: " + signature.Hash.String())
+	}
+
+	sigsToCheck := []*packet.Signature{signature, verifiedKey.PrimarySelfSignature}
 	if !verifiedKey.IsPrimary() {
 		sigsToCheck = append(sigsToCheck, verifiedKey.SelfSignature, verifiedKey.SelfSignature.EmbeddedSignature)
 	}
+	var errs []error
 	for _, sig := range sigsToCheck {
-		for _, notation := range sig.Notations {
-			if notation.IsCritical && !config.KnownNotation(notation.Name) {
-				return errors.SignatureError("unknown critical notation: " + notation.Name)
-			}
+		var pk *packet.PublicKey
+		if sig == verifiedKey.PrimarySelfSignature || sig == verifiedKey.SelfSignature {
```

Comment:
```suggestion
		if sig == verifiedKey.PrimarySelfSignature {
```
Maybe I missed something here but `verifiedKey.SelfSignature` is only in the `sigsToCheck` if we deal with a subkey. 
Thus, if `sig == verifiedKey.SelfSignature` we should check with the subkey, right?

Reply:
No, because `verifiedKey.SelfSignature` is signed by the primary key, so we should check with that - although it's true that this check is only relevant when the message signature is signed by a subkey.

Reply:
Ah right, I missed this 👍